### ISSUE TITLE
Cannot tell which pane is focused if there are 2 panes and `pane_frame` is disabled

### ISSUE THREAD

**Issue by imsuck** [zellij-org/zellij#2180]
In tmux if there are 2 panes in a tab, then the pane border will be split into two. One with the inactive pane color and one with active pane color.
But on zellij it's just the active pane color. This makes it impossible to tell which pane you have focused unless if you have `pane_frame` enabled.

https://user-images.githubusercontent.com/49095435/219952668-291490c6-8a35-47be-9ef5-d01656d078fc.mp4

**Comment by imsnif** [comment]
How do you know if up is right or left? This always bothered me in tmux.

**Comment by imsuck** [comment]
> How do you know if up is right or left? This always bothered me in tmux.

If the upper pane is focused, the active pane color would be on the left half.

https://user-images.githubusercontent.com/49095435/220268658-10558cc2-200d-4015-a5ee-07a8bfd2827d.mp4

<br>

[This](https://github.com/tmux/tmux/blob/cdacc12ce305ad2f3e65e2a01328a988e3200b51/screen-redraw.c#L160-L165) is the code that tmux uses for handling 2-pane tabs.

**Comment by imsnif** [comment]
I know :)

I'm just not sure how I'm supposed to know this as a user without fiddling around with it (and then I know anyway).

Like, what's the UI logic here? Why up == left and not up == right?

**Comment by imsuck** [comment]
Well, I'm not quite sure about this one either.

**Comment by zolrath** [comment]
I think the logic here is supposed to be that (at least in western/LTR cultures) you read left to right, top to bottom.

So, in this case as we don't have top/bottom indicators they connect left and up (the start) with right and bottom (the end).

**Comment by zolrath** [comment]
I wonder if using ASCII characters for the line break between the two panes would be sufficient?

Something like

```sh
   [13:47] ~ top selected
 
└╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼┘
   [13:47] ~ bottom unselected


```

```sh
   [13:47] ~ top unselected
 
┌╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼╼┐
   [13:47] ~ bottom selected


```

**Comment by Drugoy** [comment]
I'd suggest something a bit more obvious.
Maybe like this
```
$                      | $ 
$                      > $ echo this pane is active
$ echo inactive pane   > this pane is active
inactive pane          > $
$                      |
```
Or even some unicode symbols for arrows: https://symbl.cc/en/collections/arrow-symbols/
If using unicode symbols might cause some problems - there are these symbols from ascii table that I believe should be universally available:
« » (these two, unfortunately, don't have similar pair for vertical indication)
├ ┤ ┴  ┬
╠ ╣ ╩ ╦
the border might contain just 1/3/5 such symbol(s) or consist entirely of them.
Or that part can be configurable.

**Comment by imsnif** [comment]
> I wonder if using ASCII characters for the line break between the two panes would be sufficient?
> 
> Something like [sic]

This is my favorite.

I think the idea with the various arrow fonts is cool (we use some of them for our UI :) ), but I'm afraid they'll be a little disruptive. Lots of people using `pane_frames false` are doing so to make the UI as minimal as possible.

Any volunteers to implement?

**Comment by Gryzor** [comment]
@imsuck where would one "start" by _attempting_ to implement this? what would be your starting point?

I am quite noob at Rust, but managed to compile the project (have to solve a terminal issue with the fonts but that's irrelevant I think).

Since it's a mid-sized project, is there any "architecture" guide that I've missed or somewhere where you can give me the "30 sec" intro to the source code?

I feel tempted to say "I can't promise much" because despite my extensive programming "experience" (many years) I am a newbie with Rust and multi-platform terminal multiplexers, I may not be the right fit for this task, but I'd love to give it a try, time permitting.

**Comment by marianoguerra** [comment]
maybe "dim" the unfocused panes? have less contrast, less contrast or a lighter/darker background?

**Comment by Gryzor** [comment]
> maybe "dim" the unfocused panes? have less contrast, less contrast or a lighter/darker background?

This introduces potential accessibility issues, as not all users may be able to distinguish them.

**Comment by cbr9** [comment]
I know it is slightly unrelated, but what if there was an option to hide all pane frames while there is only one pane in that tab, and when the user opens a new pane, then show the frames?

**Comment by Drugoy** [comment]
@cbr9  Huh? That's already how it is, unless I misunderstand you.

**Comment by cbr9** [comment]
No. As of now, if I set pane_frames = false, zellij will open with no pane frames, and if I open new panes, it will still not show pane frames. I would like an option to only hide the pane frames when there is only one pane open.

**Comment by Drugoy** [comment]
Just don't set it to false and you'll get what you want, probably.

**Comment by Gryzor** [comment]
> Just don't set it to false and you'll get what you want, probably.

Not really, what @cbr9 wants is "only show frames if there's more than one pane", an option that doesn't currently exist.

**Comment by Drugoy** [comment]
I don't have frames when there's just one pane.

**Comment by mejo13** [comment]
Hey, i'm currently trying to implement this and i was wondering whether it should be a settable option for which part of the boundarie relates to which pane?

**Comment by imsnif** [comment]
Hey @mejo13 - happy you're working on this. Maybe we can implement the idea from this comment? https://github.com/zellij-org/zellij/issues/2180#issuecomment-1439187219

I think that will be very clear and not require any configs.

**Comment by Gryzor** [comment]
Agree. I was gonna try to work on this, but if you want to, go for it @mejo13 
I don't have much time, rust knowledge, and zellij codebase knowledge to probably do it right.

The way I though approaching it was mostly detecting whether frames were on/off and changing which characters the boundaries used when frames were off. But I didn't get further than _finding_ where the boundaries were "drawn". 

If you have a public fork/branch, I'd be happy to take a look at your approach, maybe there's something I can learn ;)

**Comment by mejo13** [comment]
My bad, should have read the thread more carefully. Right now i have _something_ working but it mimics tmux behavior and needs more work anyway. Going to wrap some things up and push it to GitHub then. I could give the proposed solution from [#2180]([url](https://github.com/zellij-org/zellij/issues/2180#issuecomment-1439187219)) a try nonetheless, but it would probably take some time. Sorry for the confusion!

**Comment by voroskoi** [comment]
Hi,

I just tried zellij a couple of days ago, and I really miss something similar. I do not like wasting space on these separators and there is a better way (for my taste).
In tmux there are `window-style` and `window-active-style` options, so You can set different background for those. I usually set the active window to `black` (which is taken from the terminal, actually dark grey) and other windows to #000000.
This way the background color defines the active window/pane. I would love that in zellij too.

Thanks,

**Comment by am-zest** [comment]
> Hi,
> 
> I just tried zellij a couple of days ago, and I really miss something similar. I do not like wasting space on these separators and there is a better way (for my taste). In tmux there are `window-style` and `window-active-style` options, so You can set different background for those. I usually set the active window to `black` (which is taken from the terminal, actually dark grey) and other windows to #000000. This way the background color defines the active window/pane. I would love that in zellij too.
> 
> Thanks,

that's exactly what I do, though I set the active window to '#222222'

**Comment by am-zest** [comment]
I would vote strongly for dimming the inactive panes, similar to tmux, but I also think the point that someone made about accessibility might be valid. Ideally, multiple options would be available. I'd prioritize dimming the inactive panes. :)

**Comment by doneill** [comment]
> maybe "dim" the unfocused panes? have less contrast, less contrast or a lighter/darker background?

> I would vote strongly for dimming the inactive panes, similar to tmux, but I also think the point that someone made about accessibility might be valid. Ideally, multiple options would be available. I'd prioritize dimming the inactive panes. :)

Warp has a setting for this, it has been terribly difficult to know which pane is active especially when returning back to the terminal and start typing in the wrong pane 🤯

**Comment by eins** [comment]
> > maybe "dim" the unfocused panes? have less contrast, less contrast or a lighter/darker background?
> 
> > I would vote strongly for dimming the inactive panes, similar to tmux, but I also think the point that someone made about accessibility might be valid. Ideally, multiple options would be available. I'd prioritize dimming the inactive panes. :)
> 
> Warp has a setting for this, it has been terribly difficult to know which pane is active especially when returning back to the terminal and start typing in the wrong pane 🤯

Oh I noticed that zellij doesn't have an option to displays the pane numbers, that could be an easy way to know in which pane you left zellij. 
Having dim unfocused panes can be a problem if you want to see logs for example in the no active panes.

**Comment by tombh** [comment]
Just to confirm: Zellij doesn't currently have a way to set the active pane's background colour? We're waiting on https://github.com/zellij-org/zellij/issues/2297 for that?

**Comment by thomasschafer** [comment]
I think a way to dim unfocused panes, separate to theming, would be ideal - maybe a standalone `unfocused_pane_dim` config option which could be e.g. < 1 to dim unfocused panes (for light themes) or > 1 to brighten them (for dark themes). Here's how I have tmux set up, which I guess would correspond to a value of something like 1.2 as unfocused panes are slightly brighter.

<img width="1712" alt="Image" src="https://github.com/user-attachments/assets/15da9617-0ec1-4c73-81c4-0b6ef5dd789f" />

I'm happy to have a crack at implementing this - any thoughts or objections?

**Comment by Drugoy** [comment]
I think that'd be an overkill. I think the problem is actual only when there are 2 panes: in all other cases you can look for the pane with green border around it.
I don't know the capabilities of the terminals and I don't know if that is possible, but I'd suggest to provide the solution just for the 2 cases when there are 2 panes: either split via a vertical or a horizontal line.
My suggested solution is simply making the splitting line doubled and coloring only 1 of the 2 splitting lines with green: the one closer to the active pane.

**Comment by thomasschafer** [comment]
> I think that'd be an overkill. I think the problem is actual only when there are 2 panes: in all other cases you can look for the pane with green border around it. I don't know the capabilities of the terminals and I don't know if that is possible, but I'd suggest to provide the solution just for the 2 cases when there are 2 panes: either split via a vertical or a horizontal line. My suggested solution is simply making the splitting line doubled and coloring only 1 of the 2 splitting lines with green: the one closer to the active pane.

I disagree here - dimming unfocused panes is a fairly standard solution to this problem across other terminals and multiplexers. There may well be good solutions using the shape of the lines but that doesn’t preclude a dimming solution too

**Comment by Gryzor** [comment]
Dimming may not be accessible enough, unless there's sufficient contrast. Color alone wouldn't be necessarily enough.

**Comment by thomasschafer** [comment]
I’m advocating for dimming as an _additional_ option, independent of line shapes - I think it would be great to have both

**Comment by Gryzor** [comment]
> I’m advocating for dimming as an _additional_ option, independent of line shapes - I think it would be great to have both

Agree that's a nice thing to have.

In an ideal world we'd both have time to implement both solutions at the same time :)

**Comment by eins** [comment]
dimming and  line shapes could be distracting if you have multiple panes. A feature like  `display-panes` from tmux could be interesting because you know right away what is your active pane .

**Comment by Drugoy** [comment]
Dimming a pane feels like that's a task for a GPU-accelerated program.
I'd like my terminal to stay text.

Tmux-like `display-panes` requires an action, I'd call that solution sub-optimal.
IMO, line shape via some char - would be the optimal solution here.

**Comment by thomasschafer** [comment]
> Dimming a pane feels like that's a task for a GPU-accelerated program. I'd like my terminal to stay text.
> 
> Tmux-like `display-panes` requires an action, I'd call that solution sub-optimal. IMO, line shape via some char - would be the optimal solution here.

I disagree with your premise that there is a globally optimal solution. Lots of people in this thread have expressed their support for dimming actives panes, so for many (me included) this would be optimal. I wouldn’t need a line-shape solution, but I’m not arguing against that as a solution for others - again, I’d like to see *both*!

**Comment by ocharles** [comment]
I'm not sure it's been mentioned, but the `tmux` screenshot above seems to change the border colours for the active pane. Has that been discussed here? That probably won't clarify having exactly two panes, but any more I think it starts to become unambiguous?

Another option I'd like is to just show the pane "title bar" - I don't need all the edges, but if I could just see the title bar (which I do get when I stack panes), this would again make it super clear - I'll just bold & colour the focused title bar

**Comment by Drugoy** [comment]
My point exactly: the issue is not actual in case of more than 2 panes. For the case of just 2 panes IMO using some symbols to point to the active pane would be enough, unless there's an option to hide pane borders. In case there is such an option (already done or only planned, doesn't matter) - there needs to be another solution. People above suggested dimming.

**Comment by calamity-m** [comment]
https://github.com/zellij-org/zellij/issues/2180#issuecomment-1439187219

I'm a random passerby that uses zellij a lot - this issue is annoying to me, and this seems the best solution as opposed to performing more complex diming etc.

**Comment by imsnif** [comment]
> [#2180 (comment)](https://github.com/zellij-org/zellij/issues/2180#issuecomment-1439187219)
> 
> I'm a random passerby that uses zellij a lot - this issue is annoying to me, and this seems the best solution as opposed to performing more complex diming etc.

I experimented with this a bit and IMO it doesn't look very good and can actually increase the confusion in certain cases. The upcoming version includes a new "titles only" frame-mode that I think might make things clearer: https://github.com/zellij-org/zellij/pull/5318

**Comment by Drugoy** [comment]
IMO that's a poor solution: it would occupy an extra line in each pane.

**Comment by calamity-m** [comment]
> > [#2180 (comment)](https://github.com/zellij-org/zellij/issues/2180#issuecomment-1439187219)
> > I'm a random passerby that uses zellij a lot - this issue is annoying to me, and this seems the best solution as opposed to performing more complex diming etc.
> 
> I experimented with this a bit and IMO it doesn't look very good and can actually increase the confusion in certain cases. The upcoming version includes a new "titles only" frame-mode that I think might make things clearer: [#5318](https://github.com/zellij-org/zellij/pull/5318)

Thanks. For me this solved the issue clean enough to make me happy! I am now back on zellij as this was killing me lmao. Thanks very much.
